rfcs/20250911-mxfp-support/README.md

```diff
@@ -0,0 +1,169 @@
+  already supported by oneDNN API (namely, `s8`, `f8_e5m2`, `f8_e4m3`,
+  `f4_e2m1`).
+- block scaling with 1d groups of 32. This is supported by oneDNN
+  through `set_scales` API.
+- `e8m0` block scale datatype. This is supported in oneDNN.
+- support dynamic quantization of output. This is the missing piece
+  and main point of discussion of this RFC.
+
+## MX Dynamic quantization definition
+
+How the scales are computed is defined by the MXFP spec. For a given
+1d block of size 32 (denoted $`X`$), its scale `S` and quantized values `Q`
+are defined as:
+
+$$ S = E8M0(amax(X)) / E8M0(MAX\_DST\_DT) $$
+$$ Q = dst\_dt(X / S) $$
+
+With `E8M0` a conversion to `E8M0` datatype, by rounding down to the
+closest power of two below the argument value, `MAX_DST_DT` the
+maximum representable value in `dst` type, and `dst_dt` a conversion
+to `dst` type.
+
+Note that the division operation happens _after_ rounding down to the
+scale datatype (`E8M0`). This allows to implement this division as a
+subtraction of the exponent fields. However, this differs from other
+dynamic scaling formulas, for example:
+- traditional int8 dynamic quantization applies the division _before_
+  conversion to scale datatype
+- cuDNN [^3] and cublas [^4] formula compute the division in f32
+  _before_ the conversion to scale datatype.
+
+## Proposals
+
+First of all, this RFC focuses on MXFP formats supports but might
+cover some generalizations to allow extension to other formats and
+recipes.  We will also assume that scales and data are not interleaved
+in memory, similarly to current quantization support.
+
+### Option 1 (Recommended): Through a new attribute (`set_dynamic_scales`)
+
+Here we propose to expose a new `dynamic_scales` attribute, similar to
+`scales` attribute except that:
+- it would be applicable only to output memory,
+- it implies that scales are computed by oneDNN, and not provided by user,
+- it implies a new output memory to store the computed scales.
+
+In order to support MXFP formats, it needs to support groups of 32,
+and a way to express along which axis the grouping occurs. It also
+needs to support `E8M0` format for scales. We propose to use the same
+semantics as with `set_scales` method:
+- grouping factor can be made explicit with mask, ndims and group_dims
+  arguments,
+- `E8M0` scale datatype can be set explicitly with a scale data_type
+  argument.
+- regarding the scale computation formula, we would not expose a knob
+  to configure it, and support only MXFP spec formula.
+
+This will result in the following new entry-point in C API:
+```C
+dnnl_status_t DNNL_API dnnl_primitive_attr_set_dynamic_scales(
+        dnnl_primitive_attr_t attr, int arg, int mask, int n dims,
+        const dnnl_dims_t group_dims, dnnl_data_type_t data_type);
+```
+
+And the associated symbol in C++ API:
+```c++
+    void set_dynamic_scales(int arg, int mask, const memory::dims &groups,
+            memory::data_type data_type = memory::data_type::f32);
```

Comment:
Is it the scale data type? Why defaults to f32? I see it's E8M0 scale described above.